### PR TITLE
Move native Timeout logic into Ruby

### PR DESCRIPTION
**Pull request by headius**
This is a first attempt to move our native Timeout into Ruby, using the same Executor logic we had before. This may be easier for the timeout gem to swallow, since we would not have to introduce an extension and -java platform gem.

This version actually passes three more tests than the native version, due to better integration with Ruby thread events and type conversions.

The sole test it fails, test_skip_rescue, requires that the propagation of the exception actually be via `throw` rather than `raise`, and leverages the fact that `raise` in the target thread will first call `exception` on the given object, which then throws instead of raising. This behavior may be difficult to support.

### CODE REVIEW

File: lib/ruby/stdlib/timeout.rb [hunk 1]
```diff
@@ -51,19 +54,147 @@ def exception(*)
...
+      if Fiber.respond_to?(:current_scheduler) && (scheduler = Fiber.current_scheduler)&.respond_to?(:timeout_after)
+        return scheduler.timeout_after(sec, klass || Error, message, &block)
+      end
+
+      current_thread = Thread.current
+      latch = JRuby::AtomicBoolean.new(false);
+
+      id = klass.nil? ? Object.new : nil;
+
+      sec_float = JRuby::RubyTime.convert_time_interval sec
```
[the remark this answers is not in the document]
Comment (headius) [reply lib/ruby/stdlib/timeout.rb:168]: Potentially, yes. I am pleased with how this worked out but it does require a lot of deep magic currently.

**Comment by headius** [comment]
This is now a higher priority due to net-protocol depending on timeout.

In 993cf1822c772051e0ef31e3202b84952952c635 @enebo has switched us to use the gem so we can keep functioning, after I failed to notice that net-protocol depends on timeout. That implementation is the worst one available, due to it spinning up a new thread for every single timeout operation. This PR's code would get incorporated into the timeout gem and used on JRuby.

A second option is the Rubinius implementation, which uses a single background thread and timeout request jobs, similar to the ScheduledThreadPoolExecutor we use in our logic: https://github.com/oracle/truffleruby/blob/master/lib/truffle/timeout.rb

A third option, which represents the most work and the most impact to the timeout gem, would be to incorporate our existing JRuby extension directly into the timeout gem, like we have for strscan or stringio. The biggest issue here is the fact that the timeout gem does not currently have a native extension.

**Comment by kares** [comment]
> A second option is the Rubinius implementation, which uses a single background thread and timeout request jobs

would be interesting to compare how the single threaded impl compares (performance wise - when there's a lot of timeout requests) to the proposed one - would be much simpler to maintain a single threaded timeout, even with some of the JRuby specifics ...

**Comment by headius** [comment]
@kares For the moment I have gone forward with including this in the timeout gem, plus some refactoring and optimization. Check out https://github.com/ruby/timeout/pull/14.

**Comment by headius** [comment]
This is moot now that timeout has a new efficient pure-Ruby implementation. We use the same gem that other Rubies use.

**Pull request closed without merging**